state/protocol/badger/validity.go

```diff
@@ -350,3 +350,38 @@ func validateClusterQC(cluster protocol.Cluster) error {
 	}
 	return nil
 }
+
+// SanityCheckConsensusNodeRootSnapshotValidity performs a sanity check to make sure root snapshot has enough history
+// to participate in consensus, we require one of next conditions to pass:
+//  1. IsSporkRootSnapshot() == true - this is a snapshot build from a first block of spork.
+//  2. snapshot.Head().Height - snapshot.Params().SporkRootBlockHeight() >= transaction_expiry_limit - such snapshot
+//     has enough history to validate collection guarantees and can safely participate in consensus
+func SanityCheckConsensusNodeRootSnapshotValidity(snapshot protocol.Snapshot) error {
+	isSporkRootSnapshot, err := protocol.IsSporkRootSnapshot(snapshot)
+	if err != nil {
+		return fmt.Errorf("could not check if root snapshot is a spork root snapshot: %w", err)
+	}
+	// condition 1 satisfied
+	if isSporkRootSnapshot {
+		return nil
+	}
+
+	// check condition 2
+	head, err := snapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not query root snapshot head: %w", err)
+	}
+
+	sporkRootBlockHeight, err := snapshot.Params().SporkRootBlockHeight()
+	if err != nil {
+		return fmt.Errorf("could not query spork root block height: %w", err)
+	}
+
+	cfg := DefaultConfig()
+	if head.Height+sporkRootBlockHeight < cfg.transactionExpiry {
+		return fmt.Errorf("invalid root snapshot length, expecting at least %d, got %d",
+			cfg.transactionExpiry, head.Height-sporkRootBlockHeight)
+	}
```

Comment:
This condition should be primarily concerned with the sealing segment length. I don't think the current check is sufficient.

The condition should be like:
```go
blocksInSpork := head.Height - sporkRootBlockHeight
if blocksInSpork < cfg. transactionExpiry {
  // the distance to spork root is less than transaction expiry, we need all blocks back to the spork root.
  if len(sealingSegment.AllBlocks()) != blocksInSpork {
    // err
  }
else {
  // the distance to spork root is more than transaction expiry, we need at least `transactionExpiry` many blocks
  if len(sealingSemgn.AllBlocks()) < cfg.transactionExpiry {
    // err
  }
}

// pass the check
```

Reply:
:facepalm: thanks for spotting this, completely misimplemented it